src/lib/storage/lz4/lz4_encoder.hpp

```diff
@@ -122,48 +137,348 @@ class LZ4Encoder : public SegmentEncoder<LZ4Encoder> {
+
+      Assert(compression_result > 0, "LZ4 stream compression failed");
+
+      // shrink the block vector to the actual size of the compressed result
+      compressed_block.resize(static_cast<size_t>(compression_result));
+      compressed_block.shrink_to_fit();
+
+      lz4_blocks.emplace_back(std::move(compressed_block));
+    }
+
+    // Finally, release the LZ4 stream memory.
+    LZ4_freeStreamHC(lz4_stream);
+  }
+
+  /**
+   * Generate a dictionary for non-strings. The zstd dictionary is intended for string data. Therefore, the samples
+   * provided to the zstd algorithm are not the values of each row but multiple values at once (since the minimum size
+   * for a sample is 8 bytes).
+   *
+   * @tparam T The data type of the value segment. This method is only called for non-string-segments.
+   * @param values All values of the segment. They are the input data to train the dictionary.
+   * @return The trained dictionary, or in the case of failure, an empty vector.
+   */
+  template <typename T>
+  pmr_vector<char> _train_dictionary(const pmr_vector<T>& values) {
+    /**
+     * The minimum sample size is 8 bytes. Since non-string data has a constant size for each value we can just set
+     * the sample size to 8 for all values.
+     */
+    const size_t min_sample_size = 8u;
+    const auto values_size = values.size() * sizeof(T);
+    const auto sample_size = std::max(sizeof(T), min_sample_size);
+    const auto num_samples = values_size / sample_size;
+    const std::vector<size_t> sample_lens(num_samples, sample_size);
+
+    /**
+     * The recommended dictionary size is about 1/100th of the size of all samples combined, but he size also has to be
+     * at least 1KB. Smaller dictionaries won't work.
+     */
+    auto max_dictionary_size = num_samples * sample_size / 100;
+    max_dictionary_size = std::max(max_dictionary_size, _minimum_dictionary_size);
+
+    auto dictionary = pmr_vector<char>{values.get_allocator()};
+    dictionary.resize(max_dictionary_size);
+
+    const auto dictionary_size = ZDICT_trainFromBuffer(dictionary.data(), max_dictionary_size, values.data(),
+                                                       sample_lens.data(), static_cast<unsigned>(sample_lens.size()));
+    // If the generation failed, then compress without a dictionary (the compression ratio will suffer).
+    if (ZDICT_isError(dictionary_size)) {
+      return pmr_vector<char>{};
+    }
+
+    DebugAssert(dictionary_size <= max_dictionary_size,
+                "Generated ZSTD dictionary in LZ4 compression is larger than "
+                "the memory allocated for it.");
+    dictionary.resize(dictionary_size);
+    dictionary.shrink_to_fit();
+
+    return dictionary;
+  }
+
+  /**
+   * When training a dictionary for strings, each row element is a sample. This is not done when training a
+   * dictionary for non-strings since zstd's dictionary is made for strings (and the samples can have different sizes).
```

Comment:
I can't follow this part. Do you mean your sample density is 100%? Also, could you try to explain the second sentence in a different way/

Reply:
The method to train a dictionary provided by zstd expects samples, which can be of varying length. We use a fixed length of samples for data types of fixed size and the size of each of string for string segments, so every string is one sample. So yes, we train on 100% of the segment.
Zstd / LZ4 are made for huge amounts of data, so segments with 100k rows are still small and should not be a too large amount of samples. All the guideline on the amount of samples we found was "a few thousand samples" if I remember correctly.